Fixes #18902 Snapping and tracing inconsistent behaviour

lbartoletti opened the pull request:
## Description
Fixes #18902

![output](https://user-images.githubusercontent.com/7521540/45958293-5d7c2d00-c017-11e8-881e-4a3e83ba6434.gif)

cc @saberraz 


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [X] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

lbartoletti committed:
Fixes #18902 Snapping and tracing inconsistent behaviour

saberraz commented:
That is brilliant, thanks @lbartoletti

3nids commented on src/core/qgstracer.cpp:
```diff
@@ -572,16 +612,24 @@ void QgsTracer::setLayers( const QList<QgsVectorLayer *> &layers )
 }
 
-void QgsTracer::setDestinationCrs( const QgsCoordinateReferenceSystem &crs, const QgsCoordinateTransformContext &context )
+void QgsTracer::setDestinationCrs( const QgsCoordinateReferenceSystem &crs, const QgsCoordinateTransformContext &transform )
```
We use named arguments in Python and I find this confusing. Can you use `transformContext`?

3nids commented on src/core/qgstracer.cpp:
```diff
@@ -572,16 +612,24 @@ void QgsTracer::setLayers( const QList<QgsVectorLayer *> &layers )
+  invalidateGraph();
+}
+
+void QgsTracer::setRenderContext( const QgsRenderContext *render )
```
same here: `renderContext`?

3nids reviewed (dismissed):
some minimal API phrasing to update for me.
sorry, should have commented rather than requesting changes.

lbartoletti committed:
Bette naming for python

wonder-sk reviewed (commented):
Thanks for looking into that!

In addition to inline code comments:
- in point locator implementation, also styleChanged() signal is followed to possibly invalidate the index
- could you please add tests for QgsTracer for this?
- snapping respects the "/qgis/digitizing/snap_invisible_feature" setting - should tracer do the same?

nyalldawson commented on src/core/qgstracer.h:
```diff
@@ -61,10 +63,15 @@ class CORE_EXPORT QgsTracer : public QObject
      * \see destinationCrs()
      */
-    void setDestinationCrs( const QgsCoordinateReferenceSystem &crs, const QgsCoordinateTransformContext &context );
+    void setDestinationCrs( const QgsCoordinateReferenceSystem &crs, const QgsCoordinateTransformContext &transformContext );
```
Isn't this an api break? (rename of context->transformContext).

nyalldawson replied:
There's an outstanding issue on the API break issue tracker about this - maybe just add the script to that?

lbartoletti committed:
Add test
Add onStyleChanged
Add if mRenderContext when necessary
Orderer renderer if and add enableInvisibleFeature

lbartoletti commented:
So, What do I do for the name ? :wink:

nyalldawson commented:
@lbartoletti There's no choice -- it has to be left as is until 4.0

lbartoletti committed:
rerename transformContext -> context

lbartoletti commented:
@nyalldawson done. Thanks.

lbartoletti commented:
ready to merge ? :wink:

m-kuhn commented on src/core/qgstracer.h:
```diff
@@ -61,11 +63,17 @@ class CORE_EXPORT QgsTracer : public QObject
 
     /**
-     * Sets the \a crs and transform \a context used for tracing.
+     * Sets the \a crs and \a transformContext used for tracing.
```
The docstring here also needs to be reverted

lbartoletti committed:
Fix docstring

lbartoletti committed:
Adds missing styleChanged connection
Remove useless if ( mRenderContext ) in onAttributeValueChanged and
onStyleChanged

lbartoletti committed:
Merge branch 'master' into bugfix_18902

m-kuhn commented:
@lbartoletti there is still an open question by @wonder-sk has this been answered in one of the hidden discussions?

Btw, there's normally no need to click the merge master into this branch button :)

m-kuhn commented:
@3nids are your requests sorted out? in this case, can you dismiss the review or approve?

lbartoletti commented:
> @lbartoletti there is still an open question by @wonder-sk has this been answered in one of the hidden discussions?

I had processed the request, but in fact it's not necessary, it was removed in my commit a091e64 `Remove useless if ( mRenderContext ) in onAttributeValueChanged and
onStyleChanged`
> 
> Btw, there's normally no need to click the merge master into this branch button :)

Yes, was to resolve a conflicts. Sorry

m-kuhn commented:
> Yes, was to resolve a conflicts. Sorry

If it was for a conflict it's fine.

@lbartoletti do you feel confident about merging this or should / can it wait for 3.4.1 so it can be tested in a couple of nightlies before it ends up in a release?

3nids commented:
for the small part I've commented (API arguments name), I'm happy.
I'll let the final go to @wonder-sk if there's need for one.

lbartoletti commented:
> > Yes, was to resolve a conflicts. Sorry
> 
> If it was for a conflict it's fine.
> 
> @lbartoletti do you feel confident about merging this or should / can it wait for 3.4.1 so it can be tested in a couple of nightlies before it ends up in a release?

There may be improvements to be made on the readability of the code; It remains to be seen. but the bug is fixed.

I tested it and I didn't have any problems, but I'm never sure of anything:)

This should be in QGIS 3.4 but can be in 3.4.1

wonder-sk commented on src/core/qgstracer.cpp:
```diff
@@ -585,6 +629,12 @@ void QgsTracer::setDestinationCrs( const QgsCoordinateReferenceSystem &crs, cons
+void QgsTracer::setRenderContext( const QgsRenderContext *renderContext )
+{
+  mRenderContext = std::unique_ptr<QgsRenderContext>( new QgsRenderContext( *renderContext ) );
+  invalidateGraph();
```
If I read the code correctly, this introduces a performance problem - every single time we need the graph it will regenerated due to this invalidation... (when asked for shortest path (that's on every mouse move) we call init() which calls configure() which _may_ invalidate the graph if something important has changed)

Please test with a larger layer where graph creation takes a bit of time to make sure that tracing is still fast...

lbartoletti replied:
> Please test with a larger layer where graph creation takes a bit of time to make sure that tracing is still fast...

For my tests (with a large dataset befor I got the warning on max limit data) I didn't know how to see any real difference.

lbartoletti commented:
We'll merge this PR tomorrow if we don't have further remarks.

lbartoletti committed:
Paul's review: code enhancement

pblottiere commented:
Thanks you for your latest changes @lbartoletti.

I'll merge your PR by the end of the week if there's no other comments in the meantime.

pblottiere commented:
Thanks @lbartoletti :)
